File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool dfs(int& course, unordered_map<int, bool>& dico_visited, unordered_map<int, vector<int>>& dico_courses) {

        if (dico_visited.find(course) != dico_visited.end()) { return false; }
        if (dico_courses.find(course) == dico_courses.end()) { return true; }


        dico_visited[course] = true;
        for (int req : dico_courses[course]) {

            if (not dfs(req, dico_visited, dico_courses)) {
                return false;
             };

        }
        dico_visited.erase(course);
        dico_courses.erase(course);
        return true;
    };
    
    
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {

        unordered_map<int, bool> dico_visited;
        unordered_map<int, vector<int>> dico_courses;
        
        for (vector<int> req : prerequisites) {

            if (dico_courses.find(req[0]) == dico_courses.end()) {
                dico_courses[req[0]] = vector<int>{};
            }
            dico_courses[req[0]].push_back(req[1]);
        }
        
        
        
        for (vector<int> req: prerequisites) {
            if (not dfs(req[0], dico_visited, dico_courses)) {
                return false;
            }
        }

        return true;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    
    
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        dico_courses = {}
        dico_visited = {}
        for pre_req in prerequisites:
            if dico_courses.get(pre_req[0]) is None:
                dico_courses[pre_req[0]] = []
            dico_courses[pre_req[0]] += [pre_req[1]]
        
        
        
        def dfs(course):
            
            if dico_visited.get(course) is not None:
                return False
            if dico_courses.get(course) is None:
                return True
            
            dico_visited[course] = True
            for req in dico_courses[course]:
                if not dfs(req):
                    return False
            dico_visited[course] = None
            dico_courses[course] = None
            return True
        
        for pre_req in prerequisites:
            if not dfs(pre_req[0]):
                return False
            
        
        return True
        
        
        